### PR TITLE
[async] Optimizations on StateFlowGraph

### PR DESCRIPTION
**Pull request by yuanming-hu**
Related issue = #742

Sorry about the big change. I have to do this to meet a research milestone presentation at MIT (to keep my Ph.D. student status healthy...). I list the items here so that people don't reinvent the wheels. Although there are quite a few different things, each item is not huge:

1. `StateFlowGraph::Node` now has a reference to `TaskMeta` (instead of storing some meta info on its own)
2. List generation elimination (`StateFlowGraph::optimize_listgen`)
3. Dead `ClearListStmt` elimination (`StateFlowGraph::eliminate_dead_store`)
4. Empty task removal (`StateFlowGraph::eliminate_dead_store`)
5. Added `StateFlowGraph::replace_reference`
6. Added `StateFlowGraph::delete_nodes`
7. Added `StateFlowGraph::reid_nodes`
8. Added `StateFlowGraph::topo_sort_nodes`
9. Added `StateFlowGraph::disconnect_all/with`
10. `CompileConfig::async_opt_fusion/async_opt_listgen/async_opt_activation_demotion/async_opt_dse` for controlling async optimizations.
11.  `std::string CompileConfig::async_opt_intermediate_file`: when non-empty, AsyncEngine will dump each stage of optimization in PDF files. See `visualize_state_flow_graph.py` and `AsyncEngine::debug_sfg`. You need to have `dot` CLI from Graphviz. Also note that this is tested on Linux only.
12. `IRNode *` can be dynamic cast even if it's const.
13. `get_task_meta` for computing `TaskMeta` (results cached)
14. `TaskMeta::print` for printing & debugging task meta data.

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: taichi/program/state_flow_graph.cpp [hunk 1]
```diff
@@ -427,9 +514,212 @@ std::string StateFlowGraph::dump_dot(
...
+          node_c->input_edges[edges.first].end()) {
+        node_c->input_edges[edges.first].erase(node_a);
+
+        node_c->input_edges[edges.first].insert(node_b);
+        node_b->output_edges[edges.first].insert(node_c);
+      }
+    }
+  }
+  node_a->output_edges.clear();
+}
```
Comment (xumingkuan) [inline taichi/program/state_flow_graph.cpp:591]: If I understand correctly what this function does, it should also replace all edge *from* node A with new ones from node B?

**Review by xumingkuan — APPROVED** [review]
Cool!

File: taichi/program/state_flow_graph.cpp (hunk 1 above)
Comment (xumingkuan) [reply taichi/program/state_flow_graph.cpp:591]: Should be fixed in #1872.

Comment (yuanming-hu) [reply taichi/program/state_flow_graph.cpp:591]: Thanks!

**Pull request merged**